GitExtUtils/GitUI/Theming/TabControlPaintContext.cs

```diff
@@ -0,0 +1,309 @@
+            _parentBackColor = GetParentBackColor(tabs);
+            _selectedIndex = tabs.SelectedIndex;
+            _tabCount = tabs.TabCount;
+            _font = tabs.Font;
+            _imageSize = tabs.ImageList?.ImageSize ?? Size.Empty;
+            _enabled = tabs.Enabled;
+
+            try
+            {
+                _tabTexts = Enumerable.Range(0, _tabCount)
+                    .Select(i => tabs.TabPages[i].Text)
+                    .ToArray();
+                _tabImages = Enumerable.Range(0, _tabCount)
+                    .Select(i => GetTabImage(tabs, i))
+                    .ToArray();
+                _tabRects = Enumerable.Range(0, _tabCount)
+                    .Select(tabs.GetTabRect)
+                    .ToArray();
+                _tabBackColors = Enumerable.Range(0, _tabCount)
+                    .Select(i => tabs.TabPages[i].BackColor)
+                    .ToArray();
+            }
+            catch (ArgumentOutOfRangeException)
+            {
+                // Workaround probable bug in .NET framework, known example:
+                // tabCtrl.GetTabRect[tabCtrl.SelectedIndex] may throw ArgumentOutOfRangeException
+                // https://github.com/gitextensions/gitextensions/pull/7213#issuecomment-554760531
+                _failed = true;
+
+                // Set these to null explicitly to satisfy nullability checking. We will always verify
+                // _failed before dereferencing these.
+                _tabTexts = null!;
+                _tabImages = null!;
+                _tabRects = null!;
+                _tabBackColors = null!;
+            }
+        }
+
+        public void Paint()
+        {
+            if (_failed)
+            {
+                return;
+            }
+
+            using var canvasBrush = new SolidBrush(_parentBackColor);
+            _graphics.FillRectangle(canvasBrush, _clipRectangle);
+
+            RenderSelectedPageBackground();
+
+            IEnumerable<int> pageIndices;
+            if (_selectedIndex.IsWithin(0, _tabCount))
+            {
+                // render tabs in pyramid order with selected on top
+                pageIndices = Enumerable.Range(0, _selectedIndex)
+                    .Concat(Enumerable.Range(_selectedIndex, _tabCount - _selectedIndex).Reverse());
+            }
+            else
+            {
+                pageIndices = Enumerable.Range(0, _tabCount);
+            }
+
+            foreach (var index in pageIndices)
+            {
+                RenderTabBackground(index);
+                RenderTabImage(index);
+                RenderTabText(index, _tabImages[index] is not null);
+            }
+        }
+
+        private void RenderTabBackground(int index)
+        {
+            using var borderPen = CreateBorderPen();
+            var outerRect = GetOuterTabRect(index);
+            _graphics.FillRectangle(GetBackgroundBrush(index), outerRect);
+
+            var points = new List<Point>(4);
```

Comment:
`stackalloc Point[4]`?

Reply:
> stackalloc Point[4]

can't do, the actual size is 2-4 depending on `index`

> `List<Point> points = new(4);`

ok